torch/csrc/utils/cuda_lazy_init.cpp

```diff
@@ -5,24 +5,37 @@
 
 #include <torch/csrc/Exceptions.h>
 #include <torch/csrc/utils/object_ptr.h>
+#include <torch/csrc/utils/swap_thread_state.h>
 
 namespace torch {
 namespace utils {
 
 void cuda_lazy_init() {
-  AutoGIL g;
-  // Protected by the GIL.  We don't use call_once because under ASAN it
-  // has a buggy implementation that deadlocks if an instance throws an
-  // exception.  In any case, call_once isn't necessary, because we
-  // have taken a lock.
-  static bool run_yet = false;
-  if (!run_yet) {
-    auto module = THPObjectPtr(PyImport_ImportModule("torch.cuda"));
-    if (!module) throw python_error();
-    auto res = THPObjectPtr(PyObject_CallMethod(module.get(), "_lazy_init", ""));
-    if (!res) throw python_error();
-    run_yet = true;
-  }
+  if (swappedThreadState) {
+    // Don't acquired GIL because it is already held by this thread
+    static bool run_yet = false;
+    if (!run_yet) {
+      auto module = THPObjectPtr(PyImport_ImportModule("torch.cuda"));
+      if (!module) throw python_error();
+      auto res = THPObjectPtr(PyObject_CallMethod(module.get(), "_lazy_init", ""));
+      if (!res) throw python_error();
+      run_yet = true;
+    }
+  } else {
+    AutoGIL g;
+    // Protected by the GIL.  We don't use call_once because under ASAN it
+    // has a buggy implementation that deadlocks if an instance throws an
+    // exception.  In any case, call_once isn't necessary, because we
+    // have taken a lock.
+    static bool run_yet = false;
```

Comment:
Such copy-paste is error prone in future. Imagine if someone will change one branch and forgot to change another.
Consider moving code to something like `unsafe_cuda_lazy_init()` and calling it on both sites.